File: java_version/bfs/clone_graph.java
package bfs;

import java.util.*;

public class clone_graph {
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        if (node == null) {
            return null;
        }

        UndirectedGraphNode root = new UndirectedGraphNode(node.label);
        Map<UndirectedGraphNode, UndirectedGraphNode> map = new HashMap<>();
        Set<UndirectedGraphNode> set = new HashSet<>();
        Queue<UndirectedGraphNode> queue = new LinkedList<>();
        queue.add(node);
        set.add(node);
        map.put(node, root);
        //store the relationship node - newnode with same label-- set contains all node
        while(!queue.isEmpty()){
            UndirectedGraphNode new_node = queue.poll();
            for(UndirectedGraphNode neighboor: new_node.neighbors){
                if(set.contains(neighboor)){
                    continue;
                }
                set.add(neighboor);
                queue.add(neighboor);
                map.put(neighboor, new UndirectedGraphNode(neighboor.label));
            }
        }
        for(UndirectedGraphNode undirectedGraphNode:set){
            UndirectedGraphNode new_node  = map.get(undirectedGraphNode);
            for(UndirectedGraphNode neighboor: undirectedGraphNode.neighbors){
                new_node.neighbors.add(map.get(neighboor));
            }
        }

        return map.get(node);


    }
}

File: python_version/tag/google/cloneGraph.py
# Definition for a undirected graph node
class UndirectedGraphNode:
    def __init__(self, x):
        self.label = x
        self.neighbors = []

class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        if node is None:
            return node

        queue = []
        visited = set()
        queue.append(node)
        original_new = dict()
        original_new[node] = UndirectedGraphNode(node.label)
        visited.add(node)
        while len(queue) > 0:
            node = queue.pop(0)
            for i in node.neighbors:
                if i not in visited:
                    visited.add(i)
                    queue.append(i)
                    original_new[i] = UndirectedGraphNode(i.label)

        for node in visited:
            for i in node.neighbors:
                original_new[node].append(original_new[i])

        return original_new[node]









